Add ability to create multiple console sessions

dhruvisompura opened the pull request:
Addresses #6308, #6450

Allow users to create multiple console sessions for any registered runtimes. The inner workings of the `runtimeSessionService` have been updated to support starting multiple sessions of the same runtime. The `positronConsoleService` has been updated to support multiple console instances in a sane manner, specifically we no longer allow console instances to be re-used when starting a session unless the session was power-cycled.

There are some limitations with this change when it comes to supporting multiple sessions:
- The current change limits creating a console session for runtime X to one at a time. For example, if a user spams the "+" button, only one session for the active runtime is created at a time.
- If a user restarts a session for runtime X and attempts to start a new runtime for session X back to back, the first action will be the one to complete and the other will be ignored. This is due to us restricting start/restart to one session per runtime.

The runtime session service behavior for starting and restarting a session are intertwined but need to be teased apart for multiple console sessions. This work can be done in a follow up PR once some UX questions around start/shutdown/restart have been answered.

### Release Notes

#### New Features

- N/A

#### Bug Fixes

- N/A


### QA Notes

@:sessions @:console @:web

### Screenshots

https://github.com/user-attachments/assets/b4545c51-f55b-433f-be38-715985cdb22c

https://github.com/user-attachments/assets/d507a601-4506-436d-84dc-46bdbc43e025

dhruvisompura committed:
Add create console session button

dhruvisompura committed:
Add bare minimum ability to enable multiple console session creation

dhruvisompura committed:
Initial attempt to add buttons to console toolbar

dhruvisompura committed:
Update startingConsoles map

dhruvisompura committed:
Update session tracking to use runtimeId

dhruvisompura committed:
Show new interpreter selection menu from welcome page and empty console

dhruvisompura committed:
Remove actions temporarily

dhruvisompura committed:
Remove unused imports

dhruvisompura committed:
Update verbiage in empty console

dhruvisompura committed:
Update non-multi session supported unit test

dhruvisompura committed:
Update positron console service to support multiple console sessions

dhruvisompura committed:
Condense if statement

dhruvisompura committed:
Fix some duplicate console session bugs

dhruvisompura committed:
Track all runtime console instances

dhruvisompura commented on src/vs/workbench/services/positronConsole/browser/positronConsoleService.ts:
```diff
@@ -247,30 +256,73 @@ export class PositronConsoleService extends Disposable implements IPositronConso
 			} else {
-				// New runtime with a new language, so start a new Positron console instance.
-				this.startPositronConsoleInstance(e.session, attachMode, e.activate);
+				/**
+				 * Reuse an instance for the same runtime if we have one.
+				 *
+				 * NOTE: This logic for re-using a console instance has issues in the scenario where there
+				 * are multiple console sessions which have been shutdown/disconnected.
+				 *
+				 * If a user attempts to start up a session for a specific console instance there is no gaurantee
+				 * the console instance we attach the session to is the one used for the dead session.
+				 *
+				 * TODO: @dhruvisompura
+				 * Add a separate listener to handle this power-cycle scenario that returns the old session id
+				 * for the console instance we want to use. This allows us to find the console instance that needs to be replaced.
+				 */
```
FYI, this service needs some additional TLC for the power-cycle use case but I won't be addressing that in this PR because we plan to hide the shutdown button.

dhruvisompura replied:
I think this is the path a disconnected session would also take when reconnecting - so trying to check for a console instance with a matching session id is valid. Should probably mention that in the comment!

dhruvisompura committed:
Remove extra whitespace

dhruvisompura commented on src/vs/workbench/services/positronConsole/browser/positronConsoleService.ts:
```diff
@@ -403,14 +457,28 @@ export class PositronConsoleService extends Disposable implements IPositronConso
+		let positronConsoleInstance: PositronConsoleInstance | undefined;
+		if (!multiSessionsEnabled) {
+			positronConsoleInstance = this._positronConsoleInstancesByLanguageId.get(languageId);
+		} else {
+			if (this._activePositronConsoleInstance?.session.runtimeMetadata.languageId === languageId) {
+				// Return the active console instance for the language if there is one
+				positronConsoleInstance = this._positronConsoleInstancesBySessionId.get(this._activePositronConsoleInstance?.session.sessionId);
+			} else {
+				// Otherwise find a session for the languageId that is ready to use
+				positronConsoleInstance = Array.from(this._positronConsoleInstancesBySessionId.values()).find(consoleInstance => {
+					return consoleInstance.session.runtimeMetadata.languageId === languageId && consoleInstance.state === PositronConsoleState.Ready;
+				});
+			}
+		}
```
Now that we have many sessions to choose from when we get a request to execute code - we need to decide which of the sessions we can send the request to.

This implementation is pretty naive and we'll want to make it a bit smarter but this should be good enough for a quick first pass. 

If the active session for the user is for the language this executeCode request is for, we use its associated console instance. There's no check to see if the console instance is ready since we aren't doing anything like that in the old world. 

In the case where we need to find a different console instance because the language isn't the same for the active session we try to find a session that is ready for the language. I'm not presorting the instances list here since its a naive solution. I'm a bit worried about the inefficiency of sorting every time a user requests some code execution so left it out.

jmcphers replied:
If there's only (say) 15 of these max, sorting should be practically instantaneous so not worth worrying about IMO. But this scenario should be pretty rare in practice.

dhruvisompura commented on src/vs/workbench/services/positronConsole/browser/positronConsoleService.ts:
```diff
@@ -454,18 +522,31 @@ export class PositronConsoleService extends Disposable implements IPositronConso
+				positronConsoleInstance
+			);
+		} else {
+			// Add the Positron console instance.
+			const positronConsoleInstancesForRuntime =
+				this._positronConsoleInstancesByRuntimeId.get(session.runtimeMetadata.runtimeId) || [];
+			positronConsoleInstancesForRuntime.push(positronConsoleInstance);
+			this._positronConsoleInstancesByRuntimeId.set(
+				session.runtimeMetadata.runtimeId,
+				positronConsoleInstancesForRuntime
+			);
```
Just a quick note that `_positronConsoleInstancesByRuntimeId` and `_positronConsoleInstancesBySessionId` will be culling console instance objects as part of https://github.com/posit-dev/positron/issues/6311.

dhruvisompura commented on src/vs/workbench/services/runtimeSession/common/runtimeSession.ts:
```diff
@@ -65,13 +76,19 @@ export class RuntimeSessionService extends Disposable implements IRuntimeSession
+	// A map of the starting consoles. This is keyed by the runtimeId
+	// (metadata.runtimeId) of the runtime owning the session.
+	// This is the replacement for _startingConsolesByLanguageId for multiple console sessions
+	private readonly _startingConsolesByRuntimeId = new Map<string, ILanguageRuntimeMetadata>();
```
I want to point out that the start/restart paths in the multiple console sessions world limit us to only starting or restarting a single runtime session at a time. 

I don't think this problem needs to be resolved in this PR but should be handled in a follow up. I was thinking about this and there's a couple paths for resolving this:
1. We could tease apart the start and restart logic so that we only limit users to starting a single runtime session at a time. We can use the session id/notebook uri as the key in these scenarios which lets us restart multiple sessions for the same runtime at the same time (and disconnected sessions since they have a session id). We can keep using the `runtimeMode` and `sessionMode` for sessions that are starting up and those are limited to one per runtime.
2. We provision a session id for start requests before the session actually gets created. We don't have to tease apart start/restart logic and can provision an id as needed per request and give the id out via an event to any consumers that want to know about this request (specifically a console instance). 

I'd love to hear thoughts on these solutions and any alternatives if these won't work. I'm curious if provisioning the session Id earlier in the start process has any downsides?

jmcphers replied:
The reason that most of these guards exist is that at startup there are (or at least were) plentiful codepaths that would try to start R or Python for one reason or another. We want all of those to get coalesced into a single start.  So I think relaxing the rules for "only one startup per runtime id" would lead to some regressions. 

Nothing wrong with provisioning the session ID earlier!

dhruvisompura replied:
> So I think relaxing the rules for "only one startup per runtime id" would lead to some regressions.

Okay, it sounds like we'll want to separate start and restart so we can ensure that we only allow a single runtime to start at a time while also ensuring that restarts for a session of the same runtime can be kicked off. This should prevent the scenario where we "lose" a request for a start/restart when these actions are taken back to back.

I'll create a follow up issue for the work involved for this.

dhruvisompura commented on src/vs/workbench/services/runtimeSession/common/runtimeSession.ts:
```diff
@@ -365,7 +410,7 @@ export class RuntimeSessionService extends Disposable implements IRuntimeSession
 				undefined, // No notebook URI (console session)
 				source,
-				RuntimeStartMode.Switching,
+				!multiSessionsEnabled ? RuntimeStartMode.Switching : RuntimeStartMode.Starting,
```
I don't know if `RuntimeStartMode.Switching` is relevant in the multiple console session. I couldn't think of a scenario where we would switch the runtime for a session - so this is always set to starting instead.

dhruvisompura commented on src/vs/workbench/services/runtimeSession/common/runtimeSession.ts:
```diff
@@ -300,6 +339,8 @@ export class RuntimeSessionService extends Disposable implements IRuntimeSession
 	 * @returns A promise that resolves to the session ID when the runtime is started
 	 */
 	async selectRuntime(runtimeId: string, source: string, notebookUri?: URI): Promise<void> {
```
I've left this function in here and made minimal changes to it but in the multiple console session world all it does is create a new session if the `sessionMode` is `Console`.

dhruvisompura replied:
Update: 

This function can be called by extensions. The extensions can call `$selectLanguageRuntime` which ends up calling this function here. I noticed that the python extension calls this whenever a new interpreter is started for the first time to ensure we have a session for that runtime. We were also using this function to kick off session creation via different UI gestures (welcome page, interpreter action bar, etc).

This caused some interesting race conditions where we'd get two sessions for the same runtime when the user created the first session of a runtime. The fix for this was to check if we are starting any sessions for the  runtime OR have an existing session for the runtime. This change makes it so an extension does not have the ability to create additional sessions like a user does. It can only kick of creation of a session if there was a request to start a runtime that does not have a session. The UI is no longer calling this function in the multiple console session and its solely reserved for extensions. 

This opens up some questions around how extensions will be able to interact/manage runtime sessions!

dhruvisompura commented on src/vs/workbench/services/runtimeSession/common/runtimeSession.ts:
```diff
@@ -862,17 +912,43 @@ export class RuntimeSessionService extends Disposable implements IRuntimeSession
-			return this._startingConsolesByLanguageId.size > 0 ||
-				this._consoleSessionsByLanguageId.size > 0;
+			let hasRunningConsole = false;
+			const hasRunningConsoleForLanguageId = Array.from(this._consoleSessionsByRuntimeId.values()).some((sessions) => {
+				// Check if this runtime has any console sessions
+				if (sessions.length > 0) {
+					hasRunningConsole = true;
+					// Does this runtime belong to the language we are interested in?
+					if (sessions[0].runtimeMetadata.languageId === languageId) {
+						return true;
+					}
+				}
+				return false;
+			});
+
+			if (languageId) {
+				return Array.from(this._startingConsolesByRuntimeId.values()).some(
+					runtime => runtime.languageId === languageId) || hasRunningConsoleForLanguageId;
+
+			} else {
+				return this._startingConsolesByRuntimeId.size > 0 || hasRunningConsole;
+			}
```
FYI, I know this pretty inefficient but it gets the job done for now.

jmcphers replied:
I think this is fine! These numbers involved are really small.

dhruvisompura commented on src/vs/workbench/services/runtimeSession/common/runtimeSession.ts:
```diff
@@ -952,14 +1030,35 @@ export class RuntimeSessionService extends Disposable implements IRuntimeSession
+				}
+			}
+		} else {
+			/**
+			 * TODO:
+			 * from @dhruvisompura
+			 * Can the sessions be filtered to improve efficiency of search by languageId/runtimeId?
+			 * Iterating through all the active sessions won't scale well.
+			 */
+			// Enumerate the active sessions and attempt to open the resource.
+			for (const sessions of this._consoleSessionsByRuntimeId.values()) {
+				for (const session of sessions) {
+					try {
+						if (await session.openResource(resource)) {
+							return true;
+						}
+					} catch (reason) {
+						this._logService.error(`Error opening resource "${resource.toString()}". Reason: ${reason}`);
+					}
```
Rather than checking every session we have we can probably just use the session returned from `_lastActiveConsoleSessionByLanguageId`. I think I wrote all this before I added that map.

dhruvisompura committed:
Hide power button

jmcphers commented on src/vs/workbench/services/runtimeSession/common/runtimeSession.ts:
```diff
@@ -862,17 +912,43 @@ export class RuntimeSessionService extends Disposable implements IRuntimeSession
-				this._consoleSessionsByLanguageId.has(languageId);
+		const multiSessionsEnabled = multipleConsoleSessionsFeatureEnabled(this._configurationService);
+
+		if (!multiSessionsEnabled) {
```
Super nit but I think these would be easier to read if they used positive logic:

```javascript
if (multiSessionsEnabled) {
    // some new things
 } else {
    // some old things
 }
```

dhruvisompura replied:
Happy to make that change!

jmcphers commented on src/vs/workbench/browser/parts/positronTopActionBar/components/topActionBarInterpretersManager.tsx:
```diff
@@ -22,8 +22,7 @@ import { ActionBarCommandButton } from '../../../../../platform/positronActionBa
 import { localize } from '../../../../../nls.js';
-
-const runtimePickerAction = 'workbench.action.language.runtime.openActivePicker';
+import { LANGUAGE_RUNTIME_OPEN_ACTIVE_SESSIONS_ID } from '../../../../contrib/languageRuntime/browser/languageRuntimeActions.js';
```
Do the import rules allow you to reach into `config` here? (typically you can't -- maybe this should go into another file?)

dhruvisompura replied:
Ah! That is a good point. I will double check and then fix!

dhruvisompura committed:
Notify user they have exceeded the max number of console sessions

dhruvisompura committed:
Improve readability of changes

- Add helper function to update console session map
- Invert feature flag checks for easier reading

dhruvisompura committed:
Fix broken import rule

dhruvisompura committed:
Use newest console instance for code execution

dhruvisompura committed:
Update Console API for pasting text to handle multiple consoles

dhruvisompura committed:
Fix bug causing additional console session during runtime startup

dhruvisompura committed:
Use different method to create session for UI gestures

dhruvisompura committed:
Update comment describing when to use selectRuntime

dhruvisompura committed:
Remove ability to shutdown session from test

dhruvisompura committed:
Remove todo comments

dhruvisompura commented on src/vs/workbench/api/browser/positron/mainThreadConsoleService.ts:
While doing some more testing I noticed that we had this additional service and its sibling service on the extension host side that allows the extension host access to the console to support pasting text as per https://github.com/posit-dev/positron/pull/2027.

This change probably needs some more thorough testing to ensure there aren't any regression with the changes.

dhruvisompura commented on src/vs/workbench/api/common/positron/extHostConsoleService.ts:
```diff
@@ -41,7 +46,9 @@ export class ExtHostConsoleService implements extHostProtocol.ExtHostConsoleServ
 	}
 
 	getConsoleForLanguage(id: string): positron.Console | undefined {
-		const extHostConsole = this._extHostConsolesByLanguageId.get(id);
+		// find a console for this langauge id
+		const extHostConsole = Array.from(this._extHostConsolesBySessionId.values())
+			.find(extHostConsole => extHostConsole.getLanguageId() === id);
```
We'll want to revisit this here at some point and do a better job of returning a specific console session. Currently this just returns the first console session for the requested language.

softwarenerd committed:
Fix the ConsoleInstanceInfoModalPopup so it doesn't scroll unless it needs to

softwarenerd commented:
As discussed, I pushed a change to fix the `ConsoleInstanceInfoModalPopup` so it doesn't scroll unless it needs to.

softwarenerd commented:
When I start a second Python, I am seeing:

**Error refreshing variables: RPC timed out after 5 seconds: {"jsonrpc":"2.0","method":"list"}
Error (-32603)**

![image](https://github.com/user-attachments/assets/b49dde4a-72d3-4414-a625-eb069a118982)

And the second session is not available in the Variables pane.

dhruvisompura commented:
> When I start a second Python, I am seeing:
> 
> **Error refreshing variables: RPC timed out after 5 seconds: {"jsonrpc":"2.0","method":"list"} Error (-32603)**
> 
> ![image](https://private-user-images.githubusercontent.com/853239/417811343-b49dde4a-72d3-4414-a625-eb069a118982.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDA2OTc3MjgsIm5iZiI6MTc0MDY5NzQyOCwicGF0aCI6Ii84NTMyMzkvNDE3ODExMzQzLWI0OWRkZTRhLTcyZDMtNDQxNC1hNjI1LWViMDY5YTExODk4Mi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMjI3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDIyN1QyMzAzNDhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05ZGQzODEzNmQ5ZWNkODFiMzc0NjE0NzlhNzVjNTI2YzdiMmJlNjQzNGJjNDQxYjVmMDBlNzA5ZGIxNjVmOGI0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.zZROG518HkFln7ax2pu2sowFeYC3H9p-vjTmUNPbN-M)
> 
> And the second session is not available in the Variables pane.

I'm addressing this in a follow up PR! Thanks for pointing it out here though :)

softwarenerd reviewed (commented):
I gave everything a try. I noted that when I start a second Python or R session, it does not show up in the Variables pane.

For additional Python sessions, I see an error:

<img width="594" alt="image" src="https://github.com/user-attachments/assets/5eb5501e-97a8-4526-a7b9-565b7b95e9a2" />

For additional R sessions, I do not seem to see the same error:

<img width="1551" alt="image" src="https://github.com/user-attachments/assets/7b30cb8c-103b-4a76-9bbb-7e1839de988a" />

But it's not in the Variables pane.

dhruvisompura commented on src/vs/workbench/contrib/positronConsole/browser/components/createConsoleInstanceButton.tsx:
```diff
@@ -0,0 +1,106 @@
+
+export const CreateConsoleInstanceButton = () => {
+	// Context hooks.
+	const positronConsoleContext = usePositronConsoleContext();
```
Hah! I started writing `const positronConsoleContext = usePositronConsoleContext();` because that's what I was seeing elsewhere. I will keep this in mind for the future!

softwarenerd replied:
Sometimes I am far to wordy.

dhruvisompura replied:
I prefer more verbose variable names so I personally don't mind but it can definitely get out of hand 😆

dhruvisompura committed:
Merge branch 'main' into feature/add-console-session-button

dhruvisompura commented on src/vs/workbench/contrib/positronConsole/browser/components/emptyConsole.tsx:
```diff
@@ -14,11 +14,15 @@ import { localize } from '../../../../../nls.js';
+const noSessionRunning = localize('positron.console.empty.noSessionRunning', "There is no session running.");
 const useWord = localize('positron.useWord', "Use");
 const startInterpreter = localize('positron.startInterpreter', "Start Interpreter");
+const startSession = localize('positron.console.startSession', "Start Session");
```
@juliasilge The empty console situation says the following to the user: `There is no session running. Use Start Session to start one.`

That sounds a bit clunky to me, maybe we can simplify to something like this:
1. `There is no session running. Click _here_ to start one.`
2. `There is no session running. Click _Choose Session_ to start one.` // this matches the text in the interpreter dropdown

dhruvisompura replied:
Update: I just changed this to "Start Session" in the top bar interpreter dropdown.

softwarenerd committed:
Don't draw focus rectangle on delete button on click

softwarenerd commented:
Additional little detail. I think this should say "Start Session"

![image](https://github.com/user-attachments/assets/b40977c8-5d6c-4c57-9491-dd9d03e8fa0a)

to match the Console:

![image](https://github.com/user-attachments/assets/3101b5f9-1a4f-4842-8825-328ca9710aad)

When we're in a no-session state.

dhruvisompura committed:
Update label names to match for same actions

dhruvisompura committed:
Rename context variable

softwarenerd commented:
The new Start Session button is great. I think it should simply launch into the New Session... picker:

<img width="1567" alt="image" src="https://github.com/user-attachments/assets/b88dc71d-8a6e-49be-b07a-58b05e7790b0" />

Otherwise, I have to make three clicks to start a session instead of two.

dhruvisompura committed:
Skipped session tests due to power-cycling flow

dhruvisompura committed:
Add dependency array to useEffect to prevent constant re-renders

softwarenerd reviewed (dismissed):
Looks good. I had one comment.

midleman committed:
delete test
